orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw_impl.cu

```diff
@@ -0,0 +1,222 @@
+  // by the bound of the source tensor.
+  chunk_size = min(tensor_size, chunk_start_idx + chunks.chunk_size) - chunk_start_idx;
+
+  weight_chunk_ptr = weight_tensor_ptr + chunk_start_idx;
+  grad_chunk_ptr = grad_tensor_ptr + chunk_start_idx;
+  momentum_1_chunk_ptr = momentum_1_tensor_ptr + chunk_start_idx;
+  momentum_2_chunk_ptr = momentum_2_tensor_ptr + chunk_start_idx;
+}
+
+// Torch Adam equivalence.
+template <typename T_WEIGHT, typename T_GRAD, typename T_MOMENTUM>
+__global__ void AdamWComputeMode0(
+    ChunkGroup<MTA_ADAMW_GROUP_SIZE> chunks,
+    const float alpha,
+    const float beta,
+    const float epsilon,
+    const float lr,
+    const float alpha_correction,
+    const float beta_correction,
+    const float decay) {
+  const int block_idx = blockIdx.x;
+
+  T_WEIGHT* weight_chunk_ptr;
+  T_GRAD* grad_chunk_ptr;
+  T_MOMENTUM* momentum_1_chunk_ptr;
+  T_MOMENTUM* momentum_2_chunk_ptr;
+
+  // TODO(pengwa): unroll this one for better perf.
+  int chunk_size;
+
+  PrepareMTAData(chunks, block_idx, weight_chunk_ptr, grad_chunk_ptr,
+                 momentum_1_chunk_ptr, momentum_2_chunk_ptr, chunk_size);
+
+#pragma unroll(4)
+  for (int i = threadIdx.x; i < chunk_size; i += blockDim.x) {
+    float w = static_cast<float>(weight_chunk_ptr[i]);
+    float g = static_cast<float>(grad_chunk_ptr[i]);
+    float m1 = static_cast<float>(momentum_1_chunk_ptr[i]);
+    float m2 = static_cast<float>(momentum_2_chunk_ptr[i]);
+
+    // Perform weight decay.
+    w = w - (w * lr * decay);
+
+    // Compute exponentially-averaged historical gradient.
+    m1 = alpha * m1 + (1.f - alpha) * g;
+
+    // Compute exponentially-averaged historical squared gradient.
+    m2 = beta * m2 + (1.f - beta) * g * g;
+
+    // Compute the new weight.
+    const float denom = (_Sqrt(m2) / _Sqrt(beta_correction)) + epsilon;
+    w = w - (lr * m1) / (alpha_correction * denom);
+
+    // Update the new weight and momentums.
+    weight_chunk_ptr[i] = static_cast<T_WEIGHT>(w);
+    momentum_1_chunk_ptr[i] = static_cast<T_MOMENTUM>(m1);
+    momentum_2_chunk_ptr[i] = static_cast<T_MOMENTUM>(m2);
+  }
+}
+
+// Huggingface AdamW equivalence.
+template <typename T_WEIGHT, typename T_GRAD, typename T_MOMENTUM>
+__global__ void AdamWComputeMode1(
+    ChunkGroup<MTA_ADAMW_GROUP_SIZE> chunks,
+    const float alpha,
+    const float beta,
+    const float epsilon,
+    const float lr,
+    const float lr_corrected,
+    const float decay) {
+  const int block_idx = blockIdx.x;
+
+  T_WEIGHT* weight_chunk_ptr;
+  T_GRAD* grad_chunk_ptr;
+  T_MOMENTUM* momentum_1_chunk_ptr;
+  T_MOMENTUM* momentum_2_chunk_ptr;
+  int chunk_size;
+
+  PrepareMTAData(chunks, block_idx, weight_chunk_ptr, grad_chunk_ptr,
+                 momentum_1_chunk_ptr, momentum_2_chunk_ptr, chunk_size);
+
+#pragma unroll(4)
+  for (int i = threadIdx.x; i < chunk_size; i += blockDim.x) {
+    float w = static_cast<float>(weight_chunk_ptr[i]);
```

Comment:
This is a template function, so it should use template types for w, g, m1 and m2 here